lib/stanford-mods/date_parsing.rb

```diff
@@ -85,6 +86,55 @@ def self.facet_string_for_century(date_str)
         end
       end
 
+      # get String sortable value for B.C. if we have  B.C. pattern
+      #  note that these values must *lexically* sort to create a chronological sort.
+      #  We know our data does not contain B.C. dates older than 999, so we can make them
+      #  lexically sort by subtracting 1000.  So we get:
+      #    -700 for 300 B.C., -750 for 250 B.C., -800 for 200 B.C., -801 for 199 B.C.
+      # @param [String] date_str String containing B.C.
+      # @return [String, nil] String sortable -ddd if B.C. in pattern; nil otherwise
+      def self.sortable_year_for_bc(date_str)
+        bc_matches = date_str.match(/(\d{1,4}).*B\.C\./) if date_str
+        return ($1.to_i - 1000).to_s if bc_matches
+      end
+
+      # get single facet value for B.C. if we have  B.C. pattern
+      # @param [String] date_str String containing B.C.
+      # @return [String, nil] ddd B.C.  if ddd B.C. in pattern; nil otherwise
+      def self.facet_string_for_bc(date_str)
+        bc_matches = date_str.match(/\d{1,4}.*B\.C\./) if date_str
+        return bc_matches.to_s if bc_matches
+      end
+
+      # get String sortable value from date String containing yyy, yy, y, -y, -yy, -yyy
+      #  note that these values must *lexically* sort to create a chronological sort.
+      #  We know our data does not contain negative dates older than -999, so we can make them
+      #  lexically sort by subtracting 1000.  So we get:
+      #    -983 for -17, -999 for -1, 0000 for 0, 0001 for 1, 0017 for 17
+      # @param [String] date_str String containing yyy, yy, y, -y, -yy, -yyy
+      # @return [String, nil] String sortable -ddd if date_str matches pattern; nil otherwise
+      def self.sortable_year_for_early_numeric(date_str)
+        return unless date_str.match(/^\-?\d{1,3}$/)
+        if date_str.match(/^\-/)
+          # negative number becomes x - 1000 for sorting; -005 for -995
+          num = date_str[1..-1].to_i - 1000
+          return '-' + num.to_s[1..-1].rjust(3, '0')
+        else
+          return date_str.rjust(4, '0')
+        end
+      end
+
+      # get single facet value for date String containing yyy, yy, y, -y, -yy, -yyy
+      #   negative number strings will be changed to B.C. strings
+      #   positive number strings will be left padded with zeros for clarity in the facet
+      # @param [String] date_str String containing yyy, yy, y, -y, -yy, -yyy
+      def self.facet_string_for_early_numeric(date_str)
+        return unless date_str.match(/^\-?\d{1,3}$/)
```

Comment:
Same thing w/ this matching expression, I wonder if this could be moved into a method that both methods that repeat this code could use instead (and also may be more expressive).